packages/actor-extract-links-extract-tree/lib/solver.ts

```diff
@@ -0,0 +1,395 @@
+ * @param {Variable} variable - The variable targeted inside the filter expression
+ * @param {boolean} notExpression
+ * @returns {SolutionDomain} The solution domain of the whole expression
+ */
+export function recursifResolve(
+  filterExpression: Algebra.Expression,
+  domain: SolutionDomain,
+  logicOperator: LogicOperator | undefined,
+  variable: Variable,
+  notExpression: boolean,
+): SolutionDomain {
+  // We apply an or operator by default or if the domain is empty
+  if (!logicOperator || domain.isDomainEmpty()) {
+    logicOperator = LogicOperator.Or;
+  }
+
+  if (filterExpression.expressionType === Algebra.expressionTypes.TERM
+  ) {
+    if (filterExpression.term.value === 'false') {
+      domain = domain.add({ range: A_FALSE_EXPRESSION, operator: logicOperator });
+    } else if (filterExpression.term.value === 'true') {
+      domain = domain.add({ range: A_TRUE_EXPRESSION, operator: logicOperator });
+    } else {
+      throw new MisformatedFilterTermError(`The term sent is not a boolean but is this value {${filterExpression.term.value}}`);
+    }
+  } else if (
+    // If it's an array of term then we should be able to create a solver expression
+    // hence get a subdomain appendable to the current global domain with consideration
+    // to the logic operator
+    filterExpression.args[0].expressionType === Algebra.expressionTypes.TERM &&
+    filterExpression.args.length === 2
+  ) {
+    const rawOperator = filterExpression.operator;
+    const operator = filterOperatorToSparqlRelationOperator(rawOperator);
+    if (operator) {
+      const solverExpression = resolveAFilterTerm(filterExpression, operator, [], variable);
+      let solverRange: SolutionRange | undefined;
+      if (solverExpression instanceof MissMatchVariableError) {
+        solverRange = A_TRUE_EXPRESSION;
+      } else if (solverExpression instanceof Error) {
+        throw solverExpression;
+      } else {
+        solverRange = getSolutionRange(solverExpression.valueAsNumber, solverExpression.operator)!;
+      }
+      // We can distribute a not expression, so we inverse each statement
+      if (notExpression) {
+        const invertedRanges = solverRange.inverse();
+        // We first solve the new inverted expression of the form
+        // (E1 AND E2) after that we apply the original operator
+        for (const range of invertedRanges) {
+          domain = domain.add({ range, operator: logicOperator });
+        }
+      } else {
+        domain = domain.add({ range: solverRange, operator: logicOperator });
+      }
+    }
+  } else if (
+    // It is a malformed expression
+    filterExpression.args[0].expressionType === Algebra.expressionTypes.TERM &&
+    filterExpression.args.length === 1
+  ) {
+    throw new MisformatedFilterTermError(`The expression should have a variable and a value, but is {${filterExpression.args}}`);
```

Comment:
Expressions with a single term are valid in SPARQL AFAIK.

Reply:
That's because in my case I want a boolean expression so it need to have a variable and a literal, or to simply be a boolean directly. I can clarify the comment.